Easy local backend env var

hassoncs opened the pull request:
This PR builds off my original one, but adds a nice api so the code around most urls looks similar to before:
https://github.com/Kilo-Org/kilocode/pull/3055

With this change, it's super easy to point a local Kilo development extension at a local backend just by changing the run target. This PR does _not_ remove support for the kilocodeToken approach to changing the backendUrl– that's useful for pointing the production extension at local backends

hassoncs committed:
feat: add support for local backend URL override

Add KILOCODE_BACKEND_BASE_URL environment variable to enable testing against a local Kilo Code backend. Refactor hardcoded API URLs across providers, webview, and CLI to use getKiloUrl() and getKiloUrlFromToken() functions for dynamic URL resolution. Include new launch configuration and documentation updates for local development workflow.

hassoncs committed:
Clean and fixes

hassoncs committed:
more cleanup

hassoncs committed:
Flip the fn params

pandemicsyn reviewed (approved):
lgtm. My only question (more for my knowledge), our telemetry client doesn't support this or the o.g. method. Guessing thats because we don't want to accidentally send events while cruising around during development:

 https://github.com/Kilo-Org/kilocode/blob/03b6cd1b82871e2a75ebaf8aee9dd22b11483b7a/packages/telemetry/src/PostHogTelemetryClient.ts#L132

Is there a scenario where you'd specify a _development_ KILOCODE_POSTHOG_API_KEY and then still want to use a local backend?

chrarnoldus commented on src/shared/kilocode/url.ts:
```diff
@@ -0,0 +1,58 @@
+ * In development: maps subdomains to paths (api.kilocode.ai → localhost:3000/api)
+ * In production: preserves subdomain structure (api.kilocode.ai → api.kilocode.ai)
+ */
+export function getKiloUrl(targetUrl: string = "https://kilocode.ai"): string {
```
Can we avoid the hostname parsing by having multiple functions?

```
getApiUrl(relativePath: string)
getAppUrl(relativePath: string)
```

hassoncs committed:
refactor(shared): replace getKiloUrl with getApiUrl and getAppUrl

Replace the generic getKiloUrl function with two specific functions:
- getApiUrl for API endpoints (e.g., /extension-config.json)
- getAppUrl for app/web URLs (e.g., /profile, /support)

This improves type safety and clarity by separating concerns between API and app URL generation.

BREAKING CHANGE: getKiloUrl function removed, use getApiUrl or getAppUrl instead

hassoncs committed:
refactor(shared): centralize kilocode URL utilities in @roo-code/types

Move getApiUrl, getAppUrl, and getKiloUrlFromToken functions from src/shared/kilocode/ to packages/types/src/kilocode/kilocode.ts and update all imports across the codebase. This consolidates shared utilities into the types package, reducing duplication and improving maintainability.

hassoncs committed:
refactor(types): add getExtensionConfigUrl and refactor URL utilities

- Introduce getExtensionConfigUrl function for extension config endpoint
- Refactor getApiUrl and getAppUrl to use shared buildUrl helper
- Update tests to cover new function and maintain backwards compatibility
- Use getExtensionConfigUrl in Gemini CLI provider instead of hardcoded URL

chrarnoldus commented on packages/types/src/kilocode/kilocode.ts:
```diff
@@ -56,6 +58,94 @@ export function getKiloBaseUriFromToken(kilocodeToken?: string) {
+function getGlobalKilocodeBackendUrl(): string {
+	return (
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		(typeof window !== "undefined" ? (window as any).KILOCODE_BACKEND_BASE_URL : undefined) ||
```
I think there's a way to extent a type in TypeScript

hassoncs committed:
Address PR comments

hassoncs committed:
Merge remote-tracking branch 'origin/main' into easy-local-backend-env-var

hassoncs committed:
Cleanup and revert post main-merge

hassoncs committed:
Temporarily disable flakey autocomplete tests

hassoncs committed:
ci: prevent API rate limit exceeded for actions/setup-java@v4

Add check-latest: false and token configuration to setup-java steps
in code-qa.yml and marketplace-publish.yml.

hassoncs committed:
'chore: WIP'
